### PR TITLE
Add notebook tests

### PR DESCRIPTION
**Pull request by nkorinek**
Added tests for the notebooks module, and fixed small issue found in `remove_comments` function while writing tests.

### CODE REVIEW

File: matplotcheck/tests/test_notebook.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+import pytest
+import matplotlib.pyplot as plt
+from matplotcheck.base import PlotTester
+import matplotcheck.notebook as nb
+
+
+@pytest.fixture
+def locals_dictionary_good():
+    """Abriged locals dictionary of a jupyter notebook with import in the proper
```
Comment (lwasser) [inline matplotcheck/tests/test_notebook.py:12]: @nkorinek i am not sure what this docstring means. what is this testing? what is a locals_dictionary?

File: matplotcheck/tests/test_notebook.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+        "_iii": "",
+        "_i1": "import os",
+        "_i2": 'print("Good notebook")',
+        "_i3": "locals()",
+    }
+
+
+@pytest.fixture
+def locals_dictionary_bad():
+    """Abriged locals dictionary of a jupyter notebook with import in the wrong
```
Comment (lwasser) [inline matplotcheck/tests/test_notebook.py:34]: same question here.

File: matplotcheck/tests/test_notebook.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+        "_i2": "import os",
+        "_i3": "locals()",
+    }
+
+
+def test_notebook_convert_single_axes(basic_polygon_gdf):
+    """Test convert single axes."""
+    fig, ax = plt.subplots()
+    basic_polygon_gdf.plot(ax=ax)
+    store = nb.convert_axes(plt)
```
Comment (lwasser) [inline matplotcheck/tests/test_notebook.py]: in this case - what does conver_axes return? is it just the axes object? of the figure?

i think we need a more expressive variable name here because this test file is a bit confusing and so having really good docstrings and variable names will help a lot.

File: matplotcheck/tests/test_notebook.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+        "_ii": 'print("Bad notebook")',
+        "_iii": "",
+        "_i1": 'print("Bad notebook")',
+        "_i2": "import os",
+        "_i3": "locals()",
+    }
+
+
+def test_notebook_convert_single_axes(basic_polygon_gdf):
+    """Test convert single axes."""
```
Comment (lwasser) [inline matplotcheck/tests/test_notebook.py]: Test that convert_axes returns the correct matplotlib axis object?

is that what this is doing? if so let's make the docstring really clear.

File: matplotcheck/tests/test_notebook.py [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+        PlotTester(store).assert_axis_off()
+    plt.close()
+
+
+def test_notebook_convert_multi_axes(basic_polygon_gdf):
+    """Test convert multi axes."""
+    fig, (ax1, ax2) = plt.subplots(1, 2)
+    basic_polygon_gdf.plot(ax=ax1)
+    basic_polygon_gdf.plot(ax=ax2)
+    store = nb.convert_axes(plt, which_axes="all")
```
Comment (lwasser) [inline matplotcheck/tests/test_notebook.py]: same comment here about variable name. is this a list of axes in this case? or what is the output? let's make a name that fits the output of convert_Axes

File: matplotcheck/tests/test_notebook.py [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+    fig, ax = plt.subplots()
+    basic_polygon_gdf.plot(ax=ax)
+    store = nb.convert_axes(plt)
+    with pytest.raises(AssertionError, match="Axis lines are displayed on"):
+        PlotTester(store).assert_axis_off()
+    plt.close()
+
+
+def test_notebook_convert_multi_axes(basic_polygon_gdf):
+    """Test convert multi axes."""
```
Comment (lwasser) [inline matplotcheck/tests/test_notebook.py]: Test that convert_axes returns all axes if there is a plot with subplots. 

??

File: matplotcheck/tests/test_notebook.py [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+    fig, (ax1, ax2) = plt.subplots(1, 2)
+    basic_polygon_gdf.plot(ax=ax1)
+    basic_polygon_gdf.plot(ax=ax2)
+    store = nb.convert_axes(plt, which_axes="all")
+    for axes in store:
```
Comment (lwasser) [inline matplotcheck/tests/test_notebook.py]: i think we also need a test here that it returns the correct number of axes objects. so 2 in this case if convert_Axes does what i think it does.

File: matplotcheck/tests/test_notebook.py [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+        with pytest.raises(
+            AssertionError, match="Axis lines are displayed on"
```
Comment (lwasser) [inline matplotcheck/tests/test_notebook.py]: can you clarify why you tested that the axes lines were turned off here? 
is this a duck typing type of test where you assume its an axis object if we can test that the axis lines are off?

File: matplotcheck/tests/test_notebook.py [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+        ):
+            PlotTester(axes).assert_axis_off()
+    plt.close()
+
+
+def test_notebook_convert_last_axes(basic_polygon_gdf):
+    """Test convert last axes."""
```
Comment (lwasser) [inline matplotcheck/tests/test_notebook.py]: please craft clear docstrings on all of these tests!

File: matplotcheck/tests/test_notebook.py [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+    fig, (ax1, ax2) = plt.subplots(1, 2)
+    basic_polygon_gdf.plot(ax=ax2)
+    store = nb.convert_axes(plt, which_axes="last")
+    with pytest.raises(AssertionError, match="Axis lines are displayed on"):
```
Comment (lwasser) [inline matplotcheck/tests/test_notebook.py]: this test doesn't fully make sense to me. it seems like you should have a plot with two subplots. plot something on both and then check that maybe the title of the second plot is correct. so add titles to both and then make sure  you test the title in the second plot. i the title matches the title of ax2 it is working as we think it should.

It might be better to test this way to ensure we are getting the axis object that we think we are. 

id suggest this for all tests unless you just want to ensure that an axis was returned.

File: matplotcheck/tests/test_notebook.py [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+        PlotTester(store).assert_axis_off()
+    plt.close()
+
+
+def test_notebook_convert_first_axes(basic_polygon_gdf):
+    """Test convert first axes."""
```
Comment (lwasser) [inline matplotcheck/tests/test_notebook.py]: same comments here.
1. better docstring
2. test that makes sure the correct axis object is returned.
3. better variable names (more expressive)

File: matplotcheck/tests/test_notebook.py [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+    fig, (ax1, ax2) = plt.subplots(1, 2)
+    basic_polygon_gdf.plot(ax=ax1)
+    store = nb.convert_axes(plt, which_axes="first")
+    with pytest.raises(AssertionError, match="Axis lines are displayed on"):
+        PlotTester(store).assert_axis_off()
+    plt.close()
+
+
+def test_notebook_convert_axes_error(basic_polygon_gdf):
+    """Test convert axes error."""
```
Comment (lwasser) [inline matplotcheck/tests/test_notebook.py]: Incorrect parameter value for which_axes should return a value error.

File: matplotcheck/tests/test_notebook.py [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+    fig, ax = plt.subplots()
+    basic_polygon_gdf.plot(ax=ax)
+    with pytest.raises(ValueError, match="which_axes must be one of the "):
+        nb.convert_axes(plt, which_axes="error")
+    plt.close()
+
+
+def test_error_test_count_pass(capsys):
+    """Test error test prints out correct statement when passing"""
+    nb.error_test(1, 1)
```
Comment (lwasser) [inline matplotcheck/tests/test_notebook.py:175]: boy i need to look closely at the nb module. i will. i'm not sure what error_test is even for. do you know @nkorinek  or should i dig into this more.

File: matplotcheck/tests/test_notebook.py [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+    captured = capsys.readouterr()
+    assert (
+        captured.out
+        == "ERRORS TEST: FAILED! 1 of 2 Cells ran without errors\n"
+    )
+
+
+def test_remove_comments_with_comments():
+    """Test remove comments returns string without comments"""
+    test_string = """Hello\n# comments\ntest\nstring"""
```
Comment (lwasser) [inline matplotcheck/tests/test_notebook.py:192]: what does remove_comments do? how do we use it ?

File: matplotcheck/tests/test_notebook.py [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+def test_remove_comments_without_comments():
+    """Test remove comments leaves commentless string alone"""
+    test_string = """Hello\ntest\nstring"""
+    commentless = nb.remove_comments(test_string)
+    assert commentless == """Hello\ntest\nstring"""
+
+
+def test_import_test_pass(capsys, locals_dictionary_good):
+    """Test import test passes when import are done correctly"""
+    nb.import_test(locals_dictionary_good, 3)
```
Comment (lwasser) [inline matplotcheck/tests/test_notebook.py]: same question here - what is import_test for?

**Review by lwasser — CHANGES_REQUESTED** [review]
ok @nkorinek this needs some work. some of my comments are actually questions about functionality. Others are about making the tests very clear and concise with expressive variable names and docstrings that are more clear. please respond to my questions and i'll make some more time later or next week to look more closely at this module.

File: matplotcheck/tests/test_notebook.py (hunk 15 above)
Comment (lwasser) [reply matplotcheck/tests/test_notebook.py]: ok i spent some time looking at the code. there are some very cool helper functions in there that we need to document. let's work together to figure out how they are used.  it looks like import_test -- ensure that no imports exist beyond the top cell.

Can you please test this out? and note that sometimes we have markdown at the top of the notebook so let's see that it still passes with markdown cells above the imports cell!! this is cool

File: matplotcheck/tests/test_notebook.py (hunk 14 above)
Comment (lwasser) [reply matplotcheck/tests/test_notebook.py:192]: i had a look at remove comments and i'd like to test it out. i'm not sure if it's intended to try to uncomment commented out code or what it would do. we will need to test these things as implemented in a notebook i think. but we may need to chat a bit more about the functionality in this module.

**Comment by lwasser** [comment]
ok i spent some more time in this module.

1. how do we ensure that things work in a notebook? i wonder if we need to make a notebook object somehow that we can test on
2. i think we want to rename the imports function to be something more expressive
3. i'm unsure of how that comments function should work in a notebook. so let's play with it a bit. it may be looking for commented out code but i just don't know why kristen created that function and how it runs in a notebook. 

it's worth spending a bit of time on this module to better document and then understand how it works. 

don't worry @nkorinek this is really normal based on other efforts with this package. as we build tests we find issues and opportunities in the code and we make it better!

File: matplotcheck/notebook.py [hunk 16]
```diff
@@ -77,7 +77,13 @@ def remove_comments(input_string):
...
+    return "\n".join(
+        [
+            line.split("#")[0]
+            for line in split_lines
+            if len(line.split("#")[0]) != 0
+        ]
+    )
 
 
 def import_test(var_dict, n):
```
Comment (lwasser) [inline matplotcheck/notebook.py]: i think we want to rename this function. maybe something like 
`test_package_imports` 
then we may want to clean up the docstring once we understand how it works and whether there are edge cases. i do think we NEED a notebook to test all of the things in this module.

File: matplotcheck/tests/test_notebook.py (hunk 1 above)
Comment (lwasser) [reply matplotcheck/tests/test_notebook.py:12]: please add a sentence or two here about using `locals()` in a notebook! it will clarify everything! 

Abridged output of the `locals()` function which returns....

File: matplotcheck/notebook.py (hunk 16 above)
Comment (lwasser) [reply matplotcheck/notebook.py]: `test_imports_at_top`

Comment (lwasser) [reply matplotcheck/notebook.py]: and maybe make the docstring more clear?

File: matplotcheck/tests/test_notebook.py (hunk 3 above)
Comment (lwasser) [reply matplotcheck/tests/test_notebook.py]: tests
- [ ] does it return axes objects
- [ ] right number of objects are returned
- [ ] each object has the right "data" in it -- labels, or data, or title

File: matplotcheck/tests/test_notebook.py (hunk 13 above)
Comment (lwasser) [reply matplotcheck/tests/test_notebook.py:175]: 🤷 @nkorinek let's create a new issue that asks what is this test for! and we can revisit it! can you please add this issue?

File: matplotcheck/tests/test_notebook.py (hunk 14 above)
Comment (lwasser) [reply matplotcheck/tests/test_notebook.py:192]: @nkorinek let's make an issue for this one as well please. we can return to it later! Kn the issue, please be sure to document the code before vs after your edits and we can then revisit!

**Review by lwasser — CHANGES_REQUESTED** [review]
please fix the items as we discussed today in our meeting.

**Comment by nkorinek** [comment]
@lwasser this has been updated give it a look!
